### DISCUSSION TITLE
[General] Account Abstraction: Transaction Failed Error when Interacting with Contracts from Custom Wallet Implementing IAccount

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#165] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

We are attempting to interact with contracts using a custom wallet. This wallet is created by implementing the IAccount interface. One of the functionalities includes calling a contract's specific method with encoded data.

### What is the specific issue or error you're encountering?

The transaction fails when we try to interact with the contracts from our custom wallet. The error we receive is "Transaction failed" without any further details:

JS code snippet to mint nft from the custom wallet:
```
  const aaTx: types.TransactionRequest = {
    from: CUSTOM_ACCOUNT_ADDRESS,
    to: NFT_CONTRACT_ADDRESS,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(CUSTOM_ACCOUNT_ADDRESS),
    type: 113,
    customData: {
      ergsPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT
    } as types.Eip712Meta,
    value: 120000000000000, // NFT Price
    data: nftContract.interface.encodeFunctionData("mint()")
  };
```

**Error:** transaction failed [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ]

**Txn on Goerli Explorer:**
[0x0202d909969d096684cee94a273f87ee136977e24c0b2b6073e67b2a459155f5](https://goerli.explorer.zksync.io/tx/0x0202d909969d096684cee94a273f87ee136977e24c0b2b6073e67b2a459155f5)

**A similar scenario in Mainnet:**
The provided transaction hash can be used for reference: https://explorer.zksync.io/tx/0x8ec3c8b882ffc24ad804caa97a58625c9f7a58601e1447ddad8bb5bbcf2169ae

### Can you share the error messages or logs you're receiving, if any?

Error: transaction failed

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

Here's the custom account contract: https://github.com/tevaera-labs/contracts/blob/main/contracts/teva-system/SessionAccount.sol

and, below is the JS code we use to mint nft from a custom wallet:
```
  const aaTx: types.TransactionRequest = {
    from: CUSTOM_ACCOUNT_ADDRESS,
    to: NFT_CONTRACT_ADDRESS,
    chainId: (await provider.getNetwork()).chainId,
    nonce: await provider.getTransactionCount(CUSTOM_ACCOUNT_ADDRESS),
    type: 113,
    customData: {
      ergsPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT
    } as types.Eip712Meta,
    value: 120000000000000, // NFT Price
    data: nftContract.interface.encodeFunctionData("mint()")
  };

  aaTx.gasLimit = "3500000";
  aaTx.gasPrice = await provider.getGasPrice();

  const signedTxHash = EIP712Signer.getSignedDigest(aaTx);

  const signature = ethers.utils.joinSignature(
    sessionWallet._signingKey().signDigest(signedTxHash)
  );

  aaTx.customData = {
    ...aaTx.customData,
    customSignature: signature
  };

  const sentTx = await provider.sendTransaction(utils.serialize(aaTx));
```

### Have you tried to isolate the problem, and if so, what were the results?

We tried to interact with the contracts using our custom wallet, but each time we encountered the same "Transaction failed" error. We also observed that the error is consistent when calling a contract's specific method with encoded data. We haven't been provided with additional error details beyond the generic "Transaction failed" message, making it challenging to pinpoint the exact issue.

### What steps have you already taken to try to resolve the issue?

We tried multiple NFT contracts and also attempted to mint tokens using an ERC20 contract for the custom wallet. However, every time we call any method of an external contract from the custom wallet, it fails.

### Repo Link (Optional)

https://github.com/tevaera-labs/contracts/blob/main/contracts/teva-system/SessionAccount.sol

**Comment by ljankovic-txfusion** [comment, marked as answer]
Hello @gauravl-tevaeralabs,

We deployed and tested your Session Account contracts, and there seems to be an issue the way your account contract passes `msg.value `. Specifically the following lines cause the `msg.value` to be 0: 
- https://github.com/tevaera-labs/contracts/blob/b09d4c009dc4f696f902b3deb668c60ea43f14ed/contracts/teva-system/SessionAccount.sol#L86C1-L87C1  
- https://github.com/tevaera-labs/contracts/blob/b09d4c009dc4f696f902b3deb668c60ea43f14ed/contracts/teva-system/SessionAccount.sol#L93C37-L93C37

To solve this issue replace` _transaction.reserved[1]` with `_transaction.value`.

@dutterbutter If ` _transaction.reserved[1]` doesn't reference the value field of the Transaction struct, let's make sure that we update the docs.

**Reply by gauravl-tevaeralabs** [reply]
Hi @ljankovic-txfusion,
Replacing "_transaction.reserved[1]" with "_transaction.value" resolves the issue. Since this addresses the primary concern, I'm marking this as the answer and closing the discussion. However, we still have another issue that the team can address: the contract error isn't being returned for AA transactions. For instance, when I attempt to interact with an external contract from a custom account, and if a validation in the contract fails, it only returns "transaction failed". This is in contrast to a normal wallet which provides the specific contract error. The absence of the actual error message from the contract is misleading and doesn't offer a clear understanding of the failure. Thanks!

**Comment by augustinov** [comment]
It seems like the issue might be related to the transaction execution process from the custom wallet. Ensure that the transaction parameters such as nonce, gasLimit, and gasPrice are correctly set. Additionally, double-check the encoding of the function data for the contract method being called.

Since the error message is generic and doesn't provide detailed information, consider adding additional logging or error handling in your code to capture more specific error messages. This could help in diagnosing the root cause of the transaction failure.

Review the contract code for any specific conditions or requirements that need to be met when interacting with it, such as sufficient funds or permissions.

If the issue persists, consider reaching out to the provider's support or community forums for further assistance in troubleshooting the transaction failures.

**Comment by backvomeph** [comment]
Ensure that the custom wallet's implementation of the IAccount interface is correctly configured and that all necessary parameters, including gas limit and gas price, are set appropriately. Double-check the encoded function data being passed to the contract's method and verify the signature generation process. Investigate further if the contract methods require additional permissions or checks that might be causing the transaction failure.

**Comment by kennycobbs** [comment]
Given the "Transaction failed" error without additional details, it's crucial to ensure that the custom wallet implementation correctly handles the transaction signing and serialization process. Additionally, verifying that the contract method and data encoding are accurate is essential. Debugging the transaction lifecycle and checking for any potential issues with gas limits, nonce management, and contract interactions could help identify the root cause of the failure.
